File: Data_Structure/C/SinglyLinearLL.c
#include<stdio.h>
#include<stdlib.h>   

struct node
{
    int data;
    struct node *next;
};

typedef struct node NODE;
typedef struct node * PNODE;
typedef struct node ** PPNODE;

void Display(PNODE first)    
{
    while(first != NULL)       
    {
        printf("| %d | --> ",first->data);
        first = first->next;
    }

    printf("NULL\n");
   
}

int Count(PNODE first)
{
    int iCount = 0;

    while(first != NULL)       
    {
        iCount++;
        first = first->next;
    }
    
    return iCount;
}

void InsertFirst(PPNODE first, int iNo)      
{
    PNODE newn = NULL;        

    newn = (PNODE)malloc(sizeof(NODE));    

    newn->data = iNo;
    newn->next = NULL;

    if(NULL == *first)              
    {
        *first = newn;            

    }
    else                       
    {
        newn->next = *first;     
        *first = newn;
         
    }

}

void InsertLast(PPNODE first, int iNo)
{
       PNODE newn = NULL;        
       PNODE temp = NULL;          

    newn = (PNODE)malloc(sizeof(NODE));      

    newn->data = iNo;
    newn->next = NULL;

    if(*first == NULL)           
    {
        *first = newn;         

    }
    else                        
    {
        temp = *first;    

        while(temp->next != NULL)       
        {
            temp = temp->next;
        }

        temp->next = newn;
         
    }

}

void InsertAtPos(PPNODE first, int iNo, int iPos)     
{
    PNODE newn = NULL;
    PNODE temp = NULL;
    int i = 0;

    int iCount = 0;

    iCount = Count(*first);                           

    // INPUT FILTER
    if((iPos < 1) || (iPos > iCount + 1))
    {
        printf("Invalid position\n");
        return;
    }

    if(iPos == 1)
    {
        InsertFirst(first,iNo);
    }
    else if(iPos == iCount +1)
    {
        InsertLast(first,iNo);
    }
    else
    {
        newn = (PNODE)malloc(sizeof(NODE));

        newn->data = iNo;
        newn->next = NULL;

        temp = *first;

        for(i = 1; i < iPos - 1; i++)
        {
            temp = temp->next;
        }

        newn->next = temp->next;
        temp->next = newn;
    }
    
}

void DeleteFirst(PPNODE first) 
{
    PNODE temp = NULL;

    if(*first == NULL)
    {
        return;
    }
    else if ((*first)->next == NULL)
    {
        free(*first);
        *first = NULL;
    }   
    else                        
    {
        temp = *first;
        *first = (*first)->next;  
        free(temp);                
    }

}

void DeleteLast(PPNODE first)
{
    PNODE temp = NULL;

       if(*first == NULL)
    {
        return;
    }
    else if((*first)->next == NULL)       
    {
        free(*first);
        *first = NULL; 
    }
    else                           
    {
        temp = *first;

        while(temp->next->next != NULL)       
        {
            temp = temp->next;
        }

        free(temp->next);
        temp->next = NULL;
    }

}

void DeleteAtPos(PPNODE first, int iPos)           
{
    int iCount = 0;
    int i = 0;

    PNODE temp = NULL;
    PNODE target = NULL;

    iCount = Count(*first);                           

    // INPUT FILTER
    if((iPos < 1) || (iPos > iCount))
    {
        printf("Invalid position\n");
        return;
    }

    if(iPos == 1)
    {
        DeleteFirst(first);
    }
    else if(iPos == iCount)
    {
        DeleteLast(first);
    }
    else
    {
        temp = *first;

        for(i = 1; i < iPos - 1; i++)
        {
            temp = temp->next;
        }

        target = temp->next;
        temp->next = target->next;

        free(target);
    }

}

int main()
{
    PNODE head = NULL;       
    int iRet = 0;            
    int iChoice = 0;
    int iValue = 0;
    int iPosition = 0;

    while(iChoice != 9)
    {
        printf("-----------------------------------------\n");
        printf("Enter your choice : \n");
        printf("-----------------------------------------\n");

        printf("1 : Insert node at first position : \n");
        printf("2 : Insert node at last position : \n");
        printf("3 : Insert node at given position : \n");
        printf("4 : Delete node at first position : \n");
        printf("5 : Delete node at last position : \n");
        printf("6 : Delete node at given position : \n");
        printf("7 : Display elements in Linked List : \n");
        printf("8 : Number of nodes in Linked List : \n");
        printf("9 : Exit\n");
        printf("-----------------------------------------\n");

        scanf("%d",&iChoice);

        switch(iChoice)
        {
        case 1:
            printf("Enter the value : \n");
            scanf("%d",&iValue);
            InsertFirst(&head,iValue);
            break;

        case 2:
            printf("Enter the value : \n");
            scanf("%d",&iValue);
            InsertLast(&head,iValue);
            break;

        case 3:
            printf("Enter the value : \n");
            scanf("%d",&iValue);

            printf("Enter position : \n");
            scanf("%d",&iPosition);

            InsertAtPos(&head,iValue,iPosition);
            break;

        case 4:
            
            DeleteFirst(&head);
            break;

        case 5:
            
            DeleteLast(&head);
            break;

        case 6:

            printf("Enter position : \n");
            scanf("%d",&iPosition);
            
            DeleteAtPos(&head,iPosition);
            break;

        case 7:

            printf("Elements of Doubly Linked List are : \n");
            Display(head);
            break;

        case 8:

            iRet = Count(head);
            printf("Number of elements are : %d\n",iRet);
            break;

        case 9:

            printf("Thank you for using Singly Linear Linked List Application\n");
            break;
        
        default:
            printf("Invalid choice\n");
            break;
        }
    }
   
    return 0;
}

File: Data_Structure/Java/SinglyLinearLL.java
// Singly Linear Linked List

class node
{
    public int data;
    public node next;    

    node(int no)
    {
        this.data = no;
        this.next = null;
    }
}

class SinglyLL
{
    private node first;
    private int iCount;

    public SinglyLL()
    {
        this.first = null;
        this.iCount = 0;
    }

    public void Display()
    {
        node temp = first;

        while(temp != null)
        {
            System.out.print("| "+temp.data+" | - > ");
            temp = temp.next;
        }

        System.out.println("null");
    }

    public int Count()
    {
        node temp = first;
        int iCount = 0;

        while(temp != null)
        {
            iCount++;
            temp = temp.next;
        }

        return iCount;
    }

    public void InsertFirst(int iNo)
    {
        node newn = new node(iNo);

        if(first == null)
        {
            first = newn;
        }
        else
        {
            newn.next = first;
            first = newn;
        }

        iCount++;

    }   
    
    public void InsertLast(int iNo)
    {
        node newn = new node(iNo);
        //node temp = null;

        if(first == null)
        {
            first = newn;
        }
        else
        {    
            node temp = first;

            while(temp.next != null)
            {
                temp = temp.next;               
            }    
            
            temp.next = newn;
        }

        iCount++;

    }

    public void InsertAtPos(int iNo, int iPos)
    {
        node temp = null;
        node newn = null;
  
        int i = 0;

        // Input Filter

        if((iPos < 1) || (iPos > iCount + 1))
        {
            System.out.println("Invalid position");
            return;
        }

        if(iPos == 1)
        {
            InsertFirst(iNo);
        }
        else if(iPos == iCount + 1)
        {
            InsertLast(iNo);
        }

        else
        {
            newn = new node(iNo);
            temp = first;

            for(i = 1; i < (iPos - 1); i++)
            {
                temp = temp.next;
            }

            newn.next = temp.next;
            temp.next = newn;

            iCount++;
        }

    }

    public void DeleteFirst()
    {
        if(first == null)
        {
            return;
        }

        else if(first.next == null)
        {
            first = null;
        }

        else
        {
            first = first.next;
            
        }

        iCount--;

    }

    public void DeleteLast()
    {
         if(first == null)
        {
            return;
        }

        else if(first.next == null)
        {
            first = null;
        }

        else
        {
            node temp = first;

            while(temp.next.next != null)
            {
                temp = temp.next;
            }

            temp.next = null;
            
        }

        iCount--;
    }

    public void DeleteAtPos(int iPos)
    {
        node temp = null;  
        int i = 0;

        // Input Filter

        if((iPos < 1) || (iPos > iCount))
        {
            System.out.println("Invalid position");
            return;
        }

        if(iPos == 1)
        {
            DeleteFirst();
        }
        else if(iPos == iCount)
        {
            DeleteLast();
        }

        else
        {
            temp = first;

            for(i = 1; i < (iPos - 1); i++)
            {
                temp = temp.next;
            }

            temp.next = temp.next.next;

            iCount--;
        }

    }

}

public class SinglyLinearLL
{
    public static void main(String A[])
    {
        int iRet = 0;

        SinglyLL sobj = new SinglyLL();

        sobj.InsertFirst(51);
        sobj.InsertFirst(21);
        sobj.InsertFirst(11);

        sobj.InsertLast(101);
        sobj.InsertLast(111);
        sobj.InsertLast(121);


        sobj.Display();
        iRet = sobj.Count();
        System.out.println("Number of nodes in Singly Linked List : "+iRet);    
        
        sobj.DeleteFirst();
        sobj.DeleteLast();

        sobj.Display();
        iRet = sobj.Count();
        System.out.println("Number of nodes in Singly Linked List : "+iRet);  

        sobj.InsertAtPos(105,4);

        sobj.Display();
        iRet = sobj.Count();
        System.out.println("Number of nodes in Singly Linked List : "+iRet);  

        sobj.DeleteAtPos(4);

        sobj.Display();
        iRet = sobj.Count();
        System.out.println("Number of nodes in Singly Linked List : "+iRet);   

    }
}

